### PR TITLE
[CSS] @scope parsing

### PR DESCRIPTION
**Pull request by mdubet**
#### 9ff40711991962a45f64cc0c36beec81d5eb7367
<pre>
[CSS] @scope parsing
<a href="https://bugs.webkit.org/show_bug.cgi?id=264498">https://bugs.webkit.org/show_bug.cgi?id=264498</a>
<a href="https://rdar.apple.com/118180841">rdar://118180841</a>

Reviewed by Antti Koivisto.

This is boilerplate code to add initial
support for the @scope CSS rule and its parsing.

<a href="https://drafts.csswg.org/css-cascade-6/#scope-syntax">https://drafts.csswg.org/css-cascade-6/#scope-syntax</a>

It also contains a renaming of &quot;parseCSSSelector&quot; to the more exact name &quot;parseCSSSelectorList&quot;.

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/at-scope-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/at-scope-parsing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/idlharness-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-cssom-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSCSSRuleCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/contentextensions/ContentExtensionParser.cpp:
(WebCore::ContentExtensions::isValidCSSSelector):
* Source/WebCore/css/CSSPageRule.cpp:
(WebCore::CSSPageRule::setSelectorText):
* Source/WebCore/css/CSSScopeRule.cpp: Added.
(WebCore::CSSScopeRule::CSSScopeRule):
(WebCore::CSSScopeRule::create):
(WebCore::CSSScopeRule::styleRuleScope const):
(WebCore::CSSScopeRule::cssText const):
(WebCore::CSSScopeRule::start const):
(WebCore::CSSScopeRule::end const):
* Source/WebCore/css/CSSScopeRule.h: Added.
* Source/WebCore/css/CSSScopeRule.idl: Added.
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::setSelectorText):
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleBase::createCSSOMWrapper const):
(WebCore::StyleRuleBase::visitDerived):
(WebCore::StyleRuleWithNesting::StyleRuleWithNesting):
(WebCore::m_originalSelectorList):
(WebCore::StyleRuleWithNesting::create):
(WebCore::StyleRuleScope::create):
(WebCore::StyleRuleScope::copy const):
(WebCore::StyleRuleScope::StyleRuleScope):
(WebCore::StyleRuleScope::scopeStart const):
(WebCore::StyleRuleScope::scopeEnd const):
* Source/WebCore/css/StyleRule.h:
(WebCore::StyleRuleBase::isScopeRule const):
(isType):
* Source/WebCore/css/StyleRuleType.h:
* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::StyleSheetContents::traverseSubresources const):
* Source/WebCore/css/parser/CSSAtRuleID.cpp:
(WebCore::cssAtRuleID):
* Source/WebCore/css/parser/CSSAtRuleID.h:
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseSelectorList):
(WebCore::CSSParser::parseSelector): Deleted.
* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSParserEnum.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::filterProperties):
(WebCore::CSSParserImpl::parsePageSelector):
(WebCore::computeNewAllowedRules):
(WebCore::CSSParserImpl::consumeAtRule):
(WebCore::CSSParserImpl::consumeQualifiedRule):
(WebCore::CSSParserImpl::consumeImportRule):
(WebCore::fontFeatureValuesTypeMappings):
(WebCore::CSSParserImpl::consumeFontFeatureValuesRuleBlock):
(WebCore::CSSParserImpl::consumeFontFeatureValuesRule):
(WebCore::CSSParserImpl::consumeFontPaletteValuesRule):
(WebCore::CSSParserImpl::consumeKeyframesRule):
(WebCore::CSSParserImpl::consumePageRule):
(WebCore::CSSParserImpl::consumeScopeRule):
(WebCore::CSSParserImpl::consumeStyleRule):
(WebCore::CSSParserImpl::consumeBlockContent):
(WebCore::CSSParserImpl::consumeDeclarationListInNewNestingContext):
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::parseCSSSelectorList):
(WebCore::CSSSelectorParser::consumeComplexForgivingSelectorList):
(WebCore::CSSSelectorParser::consumeNestedComplexForgivingSelectorList):
(WebCore::CSSSelectorParser::consumePseudo):
(WebCore::parseCSSSelector): Deleted.
(WebCore::CSSSelectorParser::consumeForgivingComplexSelectorList): Deleted.
* Source/WebCore/css/parser/CSSSelectorParser.h:
* Source/WebCore/dom/SelectorQuery.cpp:
(WebCore::SelectorQueryCache::add):
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::flatteningStrategyForStyleRuleType):
(WebCore::isValidRuleHeaderText):
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::highlightSelector):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRule):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::selectorFilterHashCounts):

Canonical link: <a href="https://commits.webkit.org/270590@main">https://commits.webkit.org/270590@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/343bbb283965b722020022b61fa5ec3745779d5f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25809 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4416 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27090 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27906 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23629 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26094 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6177 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1848 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23737 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26058 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3324 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22237 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28486 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2942 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23193 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29262 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/22463 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23557 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23568 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27133 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/25031 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2958 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1193 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/32471 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4347 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/22942 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7056 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6215 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3413 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3275 "Built successfully") | | |

### CODE REVIEW

File: Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml [hunk 1]
```diff
@@ -1272,6 +1272,20 @@ CSSRhythmicSizingEnabled:
     WebCore:
       default: false
 
+CSSScopingStylesEnabled:
+  type: bool
+  status: preview
```
Comment (pxlcoder) [inline Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml]: `testable`? Seems a little early for `preview`.

Comment (nt1m) [reply Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml]: Agreed, we definitely do not want to enable this in STP atm

File: Source/WebCore/css/CSSScopeRule.idl [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+typedef USVString CSSOMString;
+
+[
+    Exposed=Window
+] interface CSSScopeRule : CSSGroupingRule {
```
Comment (nt1m) [inline Source/WebCore/css/CSSScopeRule.idl]:
```suggestion
[
    EnabledBySetting=CSSScopingStylesEnabled,
    Exposed=Window
] interface CSSScopeRule : CSSGroupingRule {
```

File: Source/WebCore/css/parser/CSSParserImpl.cpp [hunk 3]
```diff
@@ -478,7 +480,8 @@ RefPtr<StyleRuleCharset> CSSParserImpl::consumeCharsetRule(CSSParserTokenRange p
     return StyleRuleCharset::create();
 }
 
-enum class AllowAnonymous : bool { No, Yes };
+enum class AllowAnonymous : bool { No,
+    Yes };
```
Comment (nt1m) [inline Source/WebCore/css/parser/CSSParserImpl.cpp]: unnecessary change?

File: Source/WebCore/css/parser/CSSParserImpl.cpp [hunk 4]
```diff
@@ -936,6 +945,75 @@ RefPtr<StyleRuleCounterStyle> CSSParserImpl::consumeCounterStyleRule(CSSParserTo
...
+    // FIXME: implement mixing style nesting and scope nesting
+    if (isNestedContext())
+        return nullptr;
+
+    auto preludeRangeCopy = prelude;
+    CSSSelectorList scopeStart;
+    CSSSelectorList scopeEnd;
+    if (!prelude.atEnd()) {
+        auto successfullyConsumeScopePrelude = [&] {
+            auto successfullyConsumeScope = [&](auto& scope) {
```
Comment (nt1m) [inline Source/WebCore/css/parser/CSSParserImpl.cpp]: Not sure what the word `successfully` adds here.

**Review by anttijk — APPROVED** [review]

File: Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml [hunk 5]
```diff
@@ -1272,6 +1272,20 @@ CSSRhythmicSizingEnabled:
...
+  category: css
+  humanReadableName: "CSS Scoping Styles: @scope rule"
```
Comment (nt1m) [inline Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml]: Can we name it either `CSSScopeAtRuleEnabled` or `CSSScopingEnabled` ?

also the humanReadableName should be "CSS Scoping (@scope)" or something like that.

Comment (mdubet) [reply Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml]: CSS Scoping is a bit overlapping with Shadow DOM. Let's use the more explicit `CSSScopeAtRule`

File: Source/WebCore/css/parser/CSSSelectorParser.cpp [hunk 6]
```diff
@@ -36,6 +36,7 @@
 #include "DeprecatedGlobalSettings.h"
 #include "Document.h"
 #include "SelectorPseudoTypeMap.h"
+#include "css/parser/CSSParserEnum.h"
```
Comment (nt1m) [inline Source/WebCore/css/parser/CSSSelectorParser.cpp:39]: We usually do `#include "CSSParserEnum.h"`

File: Source/WebCore/css/parser/CSSParserImpl.cpp [hunk 7]
```diff
@@ -936,6 +944,76 @@ RefPtr<StyleRuleCounterStyle> CSSParserImpl::consumeCounterStyleRule(CSSParserTo
...
+                // Consume the left parenthesis
+                if (prelude.peek().type() != LeftParenthesisToken)
+                    return false;
+                prelude.consumeIncludingWhitespace();
+
+                // Determine the range for the selector list
+                auto selectorListRangeStart = &prelude.peek();
+                while (!prelude.atEnd() && prelude.peek().type() != RightParenthesisToken)
+                    prelude.consumeComponentValue();
+                CSSParserTokenRange selectorListRange = prelude.makeSubRange(selectorListRangeStart, &prelude.peek());
```
Comment (nt1m) [inline Source/WebCore/css/parser/CSSParserImpl.cpp:972]: `auto selectorListRange = prelude.makeSubRange(selectorListRangeStart, &prelude.peek());`
